### DISCUSSION TITLE
[U-Boot] GPIO became input after operation

### DISCUSSION THREAD

**Discussion by ltrsunny** [linux-chenxing/linux-chenxing.org#60] [General]
(Maybe issue function for the repository is necessary)
**mstar_rebase_mainline_20220409**

```
=> gpio status 54
54: output: 0 [ ]
=> gpio set 54
gpio: pin 54 (gpio 54) value is 1
=> gpio status 54
54: input: 1 [ ]
=> gpio status 55
55: output: 0 [ ]
=> gpio clear 55
gpio: pin 55 (gpio 55) value is 0
=> gpio status 55
55: input: 0 [ ]
```

IO became input after `set | clear | toggle` , that is different from what defined in [https://github.com/linux-chenxing/u-boot/blob/mstar_rebase_mainline_20220409/cmd/gpio.c](url)

**Comment by fifteenhex** [comment]
Thanks for reporting this. I will take a look

**Comment by fifteenhex** [comment]
I think it should work correctly on `mstar_rebase_mainline_20220513`. I haven't tested it yet though.

**Reply by ltrsunny** [reply]
> I think it should work correctly on `mstar_rebase_mainline_20220513`. I haven't tested it yet though.

Text changed but IO doesn't really work. Does U-boot share the same pinout with Linux? e.g. 54-GPIO12, 55-GPIO13

**Reply by fifteenhex** [reply]
Linux and u-boot are currently different because Linux has a few more pins wired up.
The u-boot numbering is based on the order in this array:
https://github.com/linux-chenxing/u-boot/blob/0fa9f56420c70842928f2b04091b32bc9deb897f/drivers/gpio/gpio-msc313.c#L150

gpio12 should be 52
gpio13 should be 53

**Reply by fifteenhex** [reply]
Just tested here and it looks right now.

**Reply by ltrsunny** [reply]
> Linux and u-boot are currently different because Linux has a few more pins wired up. The u-boot numbering is based on the order in this array: https://github.com/linux-chenxing/u-boot/blob/0fa9f56420c70842928f2b04091b32bc9deb897f/drivers/gpio/gpio-msc313.c#L150
> 
> gpio12 should be 52 gpio13 should be 53

Good, working now.
